File: resources/data-conversion/data_test.dart
import 'dart:typed_data';

import 'data.dart';

/// This will uses data from (data.dart file)
/// to perform conversation on received data.
/// the actual implementation is little bit different
/// with developer friendly way using some features provided by dart lang.,
/// the underlaying logic remains same as below.
void main() {
  const CMD_GET_APP_VERSION = 107;
  const CMD_GET_ERROR_INFORMATION = 108;
  const CMD_GET_BATTERY_VOLTAGE = 111;
  const RES_CMD_RESPONSE = 0;
  const RES_WEIGHT_MEAS = 1;
  const RES_LOW_PWR_WARNING = 4;

  int command = 0;

  for (List<int> data in dataList) {
    if (data[0] == RES_WEIGHT_MEAS) {
      // value 1 = weight measurement
      print('Payload size: ${data[1]}');
      for (int x = 2; x < data.length; x += 8) {
        // Weight
        Uint8List value = Uint8List.fromList(data.getRange(x, x + 4).toList());
        double weight = value.buffer.asByteData().getFloat32(0, Endian.little);
        // Time
        Uint8List timestamp = Uint8List.fromList(data.getRange(x + 4, x + 4 + 4).toList());
        int uSeconds = timestamp.buffer.asByteData().getUint32(0, Endian.little);
        // print
        print('$weight Kg, ${uSeconds / 1000000.0} Sec.');
      }
    } else if (data[0] == RES_LOW_PWR_WARNING) {
      print("Received low battery warning.");
    } else if (data[0] == RES_CMD_RESPONSE) {
      if (command == CMD_GET_APP_VERSION) {
        print('FW version: ${String.fromCharCodes(data.getRange(2, data.length))}');
      } else if (command == CMD_GET_BATTERY_VOLTAGE) {
        Uint8List value = Uint8List.fromList(data.getRange(2, data.length).toList());
        int voltage = value.buffer.asByteData().getUint32(0, Endian.little);
        print('Battery voltage: {$voltage} [mV]');
      } else if (command == CMD_GET_ERROR_INFORMATION) {
        print('Crashlog: ${String.fromCharCodes(data.getRange(2, data.length))}');
      }
    }
  }
}

File: Resources/data-test.py
"""
This is an extracted part (method) from the API
provided by the Tindeq team to work with progressor.
For full implementation see (Tindeq Progreessor Softwares) folder
"""

import platform
import logging
import asyncio
import struct
import csv
import time
import datetime

""" Copy and add required data lists from data.dart inside dataList = [] """
dataList = []

myData = [bytearray(d) for d in dataList]

RES_CMD_RESPONSE = 0
RES_WEIGHT_MEAS = 1
RES_RFD_PEAK = 2
RES_RFD_PEAK_SERIES = 3
RES_LOW_PWR_WARNING = 4

current_cmd_request = 1

def notification_handler(data):
    try:
        if data[0] == RES_WEIGHT_MEAS:
            # print("Payload size : {0}".format(data[1]))
            value = [data[i:i+4] for i in range (2, len(data), 8)]
            timestamp = [data[i:i+4] for i in range (6, len(data), 8)]
            for x, y in zip(value,timestamp):
                weight, = struct.unpack('<f', x) # <little endian ffloat
                useconds, = struct.unpack('<I', y) # <little endian unsigned int
                # print(weight, useconds)
        # elif data[0] == RES_LOW_PWR_WARNING:
        #     print("Received low battery warning.")

        # elif data[0] == RES_CMD_RESPONSE:

        #     if current_cmd_request == CMD_GET_APP_VERSION:
        #         print("---Device information---")
        #         print("FW version : {0}".format(data[2:].decode("utf-8")))

        #     elif current_cmd_request == CMD_GET_BATTERY_VOLTAGE:
        #         vdd, = struct.unpack('<I', data[2:])
        #         print("Battery voltage : {0} [mV]".format(vdd))

        #     elif current_cmd_request == CMD_GET_ERROR_INFORMATION:
        #         try:
        #             print("Crashlog : {0}".format(data[2:].decode("utf-8")))
        #             print("------------------------")
        #         except Exception as e:
        #             print("Empty crashlog")
        #             print("------------------------")
    except Exception as e:
        print(e)


if __name__ == "__main__":
    for i in myData:
        notification_handler(i)
